server/runtime.go

```diff
@@ -222,7 +225,35 @@ func (r *Runtime) InvokeFunctionRPC(fn *lua.LFunction, uid uuid.UUID, handle str
 	return nil, errors.New("Runtime function returned invalid data. Only allowed one return value of type String/Byte")
 }
 
-func (r *Runtime) InvokeFunctionBefore(fn *lua.LFunction, uid uuid.UUID, handle string, sessionExpiry int64, payload map[string]interface{}) (map[string]interface{}, error) {
+func (r *Runtime) InvokeFunctionBefore(fn *lua.LFunction, uid uuid.UUID, handle string, sessionExpiry int64, jsonpbMarshaler *jsonpb.Marshaler, jsonpbUnmarshaler *jsonpb.Unmarshaler, envelope *Envelope) (*Envelope, error) {
+	l, _ := r.NewStateThread()
+	defer l.Close()
+
+	ctx := NewLuaContext(l, r.luaEnv, BEFORE, uid, handle, sessionExpiry)
+	var lv lua.LValue
+	var err error
+	if envelope != nil {
+		lv, err = ConvertEnvelopeToLTable(l, jsonpbMarshaler, envelope)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	retValue, err := r.invokeFunction(l, fn, ctx, lv)
+	if err != nil {
+		return nil, err
+	}
+
+	if retValue == nil || retValue == lua.LNil {
+		return nil, nil
```

Comment:
I don't think this is/should be valid return for a Before hook. They must return the envelop or raise an error if they want to stop pipeline processing.